packages/slate/src/commands/at-range.js

```diff
@@ -807,36 +807,86 @@ Commands.insertFragmentAtRange = (editor, range, fragment) => {
   })
 }
 
-const findInsertionNode = (fragment, document, startKey) => {
-  const hasSingleNode = object => {
-    if (!object || object.object === 'text') return
-    return object.nodes.size === 1
-  }
-
-  const firstNode = object => object && object.nodes.first()
-  let node = fragment
+/**
+ * Get the deepest single child block inside `fragment` whose reversed block
+ * ancestors match the reversed block ancestors of the `document` starting at
+ * the `documentKey`.
+ *
+ * @param {Document} document
+ * @param {string} documentKey
+ * @param {Document} fragment
+ * @return {Node}
+ */
 
-  if (hasSingleNode(fragment)) {
-    let fragmentInner = firstNode(fragment)
+const findInsertionNode = (fragment, document, documentKey) => {
+  // Find the deepest block in a doc with no siblings.
+  const deepestSingleBlock = doc => {
+    let result = doc
 
-    const matches = documentNode => documentNode.type === fragmentInner.type
-    let documentInner = document.getFurthest(startKey, matches)
+    while (result.nodes.size === 1 && result.nodes.first().object === 'block') {
+      result = result.nodes.first()
+    }
 
-    if (documentInner === document.getParent(startKey)) node = fragmentInner
+    return result
+  }
 
-    while (hasSingleNode(fragmentInner) && hasSingleNode(documentInner)) {
-      fragmentInner = firstNode(fragmentInner)
-      documentInner = firstNode(documentInner)
+  // Return whether every block in the `fragmentAncestors` list has the
+  // same type as the block in `documentAncestors` with the same index.
+  const ancestorTypesMatch = (fragmentAncestors, documentAncestors) => {
+    return (
+      documentAncestors.size >= fragmentAncestors.size &&
+      fragmentAncestors.every((fragmentNode, i) => {
+        return documentAncestors.get(i).type === fragmentNode.type
+      })
+    )
+  }
 
-      if (fragmentInner.type === documentInner.type) {
-        node = fragmentInner
-      } else {
-        break
+  // Given two reverse lists of ancestors, check if all fragment ancestor types
+  // match the doc ancestors at some position.
+  const findMatchIndex = (documentAncestors, fragmentAncestors) => {
+    for (let fragIdx = 0; fragIdx < fragmentAncestors.size; fragIdx++) {
+      // The docIdx loop relaxes our check in that we can still match if there
+      // are node type differences leaf-side.
+      // This is important for example if our fragment inserts multiple siblings
+      // or inserts another type while the tree structure remains the same.
+      for (
+        let docIdx = 0;
+        docIdx <= documentAncestors.size - fragmentAncestors.size;
```

Comment:
What about the case of `fragmentAncestors.size` > `docmentAncestors.size`? If we are never getting a match in this case (which appears to be the case), we can at least return quickly.

Reply:
Agreed, just wanted to make sure that was the functionality you had in mind.

Also, since the loop will never find anything, we can probably do a check at the top, before upper `for` loop, so that we can return fast if `fragmentAncestors.size > documentAncestors.size`.

Reply:
ah, got you